File: js/packages/openinference-instrumentation-beeai/src/helpers/buildTraceTree.ts
import { OITracer } from "@arizeai/openinference-core";
import { FrameworkSpan, GeneratedResponse } from "../types";
import { SpanStatusCode, TimeInput } from "@opentelemetry/api";
import {
  OpenInferenceSpanKind,
  SemanticConventions,
} from "@arizeai/openinference-semantic-conventions";
import { startEventName, successEventName } from "../config";

interface BuiltTraceTreeProps {
  tracer: OITracer;
  mainSpanKind: OpenInferenceSpanKind;
  data: {
    prompt?: string | null;
    history: GeneratedResponse[] | undefined;
    generatedMessage: GeneratedResponse | undefined;
    spans: FrameworkSpan[];
    traceId: string;
    version: string;
    runErrorSpanKey: string;
    startTime: TimeInput;
    endTime: TimeInput;
    source: string;
  };
}

interface BuildSpansForParentProps {
  tracer: OITracer;
  data: {
    spans: FrameworkSpan[];
    traceId: string;
    parentId: string | undefined;
  };
}

function buildSpansForParent({ tracer, data }: BuildSpansForParentProps) {
  data.spans
    .filter((beeaiSpan) => beeaiSpan.parent_id === data.parentId)
    .forEach((beeaiSpan) => {
      tracer.startActiveSpan(
        beeaiSpan.context.span_id,
        {
          // custom start time
          startTime: beeaiSpan.start_time,
          // set span important attributes
          attributes: {
            target: beeaiSpan.attributes.target,
            name: beeaiSpan.name,
            traceId: data.traceId,
            ...(beeaiSpan.attributes.metadata && {
              metadata: JSON.stringify(beeaiSpan.attributes.metadata),
            }),
            ...(beeaiSpan.attributes.data && { ...beeaiSpan.attributes.data }),
          },
        },
        (activeSpan) => {
          // set status
          activeSpan.setStatus(beeaiSpan.status);

          // set nested spans
          buildSpansForParent({
            tracer,
            data: {
              spans: data.spans,
              traceId: data.traceId,
              parentId: beeaiSpan.context.span_id,
            },
          });

          // finish the span
          activeSpan.end(beeaiSpan.end_time);
        },
      );
    });
}

function buildAgentMainSpanData(data: BuiltTraceTreeProps["data"]) {
  return {
    ...(data.prompt && { [SemanticConventions.INPUT_VALUE]: data.prompt }),
    ...(data.generatedMessage !== undefined && {
      [SemanticConventions.OUTPUT_VALUE]: JSON.stringify(data.generatedMessage),
    }),
    ...(data.history && { history: JSON.stringify(data.history) }),
  };
}

function buildToolMainSpanData(data: BuiltTraceTreeProps["data"]) {
  const startBeeaiSpan = data.spans.find(
    (span) => span.name === startEventName,
  );
  const successBeeeaiSpan = data.spans.find(
    (span) => span.name === successEventName,
  );

  return {
    ...(startBeeaiSpan && {
      [SemanticConventions.INPUT_VALUE]:
        startBeeaiSpan.attributes.data?.[SemanticConventions.TOOL_PARAMETERS],
      [SemanticConventions.TOOL_PARAMETERS]:
        startBeeaiSpan.attributes.data?.[SemanticConventions.TOOL_PARAMETERS],
    }),
    ...(successBeeeaiSpan && {
      [SemanticConventions.OUTPUT_VALUE]:
        successBeeeaiSpan.attributes.data?.[SemanticConventions.OUTPUT_VALUE],
    }),
  };
}

function buildLLMMainSpanData(data: BuiltTraceTreeProps["data"]) {
  const startBeeaiSpan = data.spans.find(
    (span) => span.name === startEventName,
  );
  const successBeeeaiSpan = data.spans.find(
    (span) => span.name === successEventName,
  );

  if (!startBeeaiSpan && !successBeeeaiSpan) return {};

  const provider =
    startBeeaiSpan?.attributes.data?.[SemanticConventions.LLM_PROVIDER] ||
    successBeeeaiSpan?.attributes.data?.[SemanticConventions.LLM_PROVIDER];

  const modelName =
    startBeeaiSpan?.attributes.data?.[SemanticConventions.LLM_MODEL_NAME] ||
    successBeeeaiSpan?.attributes.data?.[SemanticConventions.LLM_MODEL_NAME];

  return {
    ...(startBeeaiSpan && {
      [SemanticConventions.INPUT_VALUE]:
        startBeeaiSpan?.attributes.data?.[SemanticConventions.INPUT_VALUE],
      [SemanticConventions.INPUT_MIME_TYPE]:
        startBeeaiSpan?.attributes.data?.[SemanticConventions.INPUT_MIME_TYPE],
    }),
    ...(successBeeeaiSpan && {
      [SemanticConventions.OUTPUT_MIME_TYPE]:
        successBeeeaiSpan.attributes.data?.[
          SemanticConventions.OUTPUT_MIME_TYPE
        ],
      [SemanticConventions.OUTPUT_VALUE]:
        successBeeeaiSpan.attributes.data?.[SemanticConventions.OUTPUT_VALUE],
    }),
    ...(provider && { [SemanticConventions.LLM_PROVIDER]: provider }),
    ...(modelName && { [SemanticConventions.LLM_MODEL_NAME]: modelName }),
  };
}

export function buildTraceTree({
  tracer,
  data,
  mainSpanKind,
}: BuiltTraceTreeProps) {
  const computedData =
    mainSpanKind === OpenInferenceSpanKind.AGENT
      ? buildAgentMainSpanData(data)
      : mainSpanKind === OpenInferenceSpanKind.TOOL
        ? buildToolMainSpanData(data)
        : mainSpanKind === OpenInferenceSpanKind.LLM
          ? buildLLMMainSpanData(data)
          : {};

  tracer.startActiveSpan(
    `beeai-framework-main`,
    {
      // custom start time
      startTime: data.startTime,
      // set main span important attributes
      attributes: {
        source: data.source,
        traceId: data.traceId,
        [SemanticConventions.OPENINFERENCE_SPAN_KIND]: mainSpanKind,
        ["beeai.version"]: data.version,
        ...computedData,
      },
    },
    (activeSpan) => {
      // set status
      const runErrorSpan = data.spans.find(
        (span) => span.attributes.target === data.runErrorSpanKey,
      );
      if (runErrorSpan) {
        activeSpan.setStatus(runErrorSpan.status);
      } else {
        activeSpan.setStatus({ code: SpanStatusCode.OK });
      }

      // set nested spans
      buildSpansForParent({
        tracer,
        data: { spans: data.spans, traceId: data.traceId, parentId: undefined },
      });

      // finish the main span with custom end time
      activeSpan.end(data.endTime);
    },
  );
}

File: python/instrumentation/openinference-instrumentation-beeai/src/openinference/instrumentation/beeai/utils/build_trace_tree.py
# Copyright 2025 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Any, List, Optional, TypedDict

from opentelemetry import trace

from openinference.instrumentation import OITracer
from openinference.semconv.trace import OpenInferenceSpanKindValues, SpanAttributes

from .create_span import FrameworkSpan


class BuildTraceTreeData(TypedDict):
    prompt: Optional[str]
    history: list[dict[str, Any]]
    generatedMessage: dict[str, Any] | None
    spans: List[FrameworkSpan]
    traceId: str
    version: str
    startTime: int
    endTime: int
    source: Optional[str]
    run_error_span_key: str


def build_spans_for_parent(
    tracer: OITracer, spans: List[FrameworkSpan], trace_id: str, parent_id: str | None
) -> None:
    children = [s for s in spans if s.get("parent_id") == parent_id]

    for span in children:
        attributes = {
            "target": span["attributes"]["target"],
            "name": span.get("name"),
            "traceId": trace_id,
        }

        if span["attributes"]["metadata"]:
            attributes["metadata"] = str(span["attributes"]["metadata"])

        if span["attributes"]["data"]:
            attributes.update(span["attributes"]["data"])

        with tracer.start_as_current_span(
            name=span["context"]["span_id"],
            attributes={k: v for k, v in attributes.items() if v is not None},
            start_time=span.get("start_time"),
        ) as current_child:
            status = span.get("status")
            if status is not None:
                current_child.set_status(status["code"])
            build_spans_for_parent(
                tracer=tracer,
                spans=spans,
                trace_id=trace_id,
                parent_id=span["context"]["span_id"],
            )


def build_agent_main_span_data(data: BuildTraceTreeData) -> dict[str, Any]:
    payload = {}
    if data.get("prompt"):
        payload[SpanAttributes.INPUT_VALUE] = data.get("prompt")
    if data.get("generatedMessage"):
        payload[SpanAttributes.OUTPUT_VALUE] = str(data.get("generatedMessage"))
    if data.get("history"):
        payload["history"] = str(data.get("history"))
    return payload


def build_tool_main_span_data(data: BuildTraceTreeData) -> dict[str, Any]:
    start_span = next((s for s in data["spans"] if s["name"] == "start"), None)
    success_span = next((s for s in data["spans"] if s["name"] == "success"), None)

    payload = {}
    if start_span and start_span["attributes"]["data"] is not None:
        tool_params = start_span["attributes"]["data"].get(SpanAttributes.TOOL_PARAMETERS)
        payload[SpanAttributes.INPUT_VALUE] = tool_params
        payload[SpanAttributes.TOOL_PARAMETERS] = tool_params
    if success_span and success_span["attributes"]["data"] is not None:
        payload[SpanAttributes.OUTPUT_VALUE] = success_span["attributes"]["data"].get(
            SpanAttributes.OUTPUT_VALUE
        )
    return payload


def build_llm_main_span_data(data: BuildTraceTreeData) -> dict[str, Any]:
    start_span = next((s for s in data["spans"] if s["name"] == "start"), None)
    success_span = next((s for s in data["spans"] if s["name"] == "success"), None)
    provider = None
    model_name = None

    if start_span and start_span["attributes"]["data"] is not None:
        provider = start_span["attributes"]["data"].get(SpanAttributes.LLM_PROVIDER)
        model_name = start_span["attributes"]["data"].get(SpanAttributes.LLM_MODEL_NAME)

    if success_span and success_span["attributes"]["data"] is not None:
        provider = provider or success_span["attributes"]["data"].get(SpanAttributes.LLM_PROVIDER)
        model_name = model_name or success_span["attributes"]["data"].get(
            SpanAttributes.LLM_MODEL_NAME
        )

    payload = {}
    if start_span and start_span["attributes"]["data"] is not None:
        payload[SpanAttributes.INPUT_VALUE] = start_span["attributes"]["data"].get(
            SpanAttributes.INPUT_VALUE
        )
        payload[SpanAttributes.INPUT_MIME_TYPE] = start_span["attributes"]["data"].get(
            SpanAttributes.INPUT_MIME_TYPE
        )
    if success_span and success_span["attributes"]["data"] is not None:
        payload[SpanAttributes.OUTPUT_VALUE] = success_span["attributes"]["data"].get(
            SpanAttributes.OUTPUT_VALUE
        )
        payload[SpanAttributes.OUTPUT_MIME_TYPE] = success_span["attributes"]["data"].get(
            SpanAttributes.OUTPUT_MIME_TYPE
        )
    if provider:
        payload[SpanAttributes.LLM_PROVIDER] = provider
    if model_name:
        payload[SpanAttributes.LLM_MODEL_NAME] = model_name

    return payload


def build_trace_tree(tracer: OITracer, main_span_kind: str, data: BuildTraceTreeData) -> None:
    if main_span_kind == OpenInferenceSpanKindValues.AGENT.value:
        computed_data = build_agent_main_span_data(data)
    elif main_span_kind == OpenInferenceSpanKindValues.TOOL.value:
        computed_data = build_tool_main_span_data(data)
    elif main_span_kind == OpenInferenceSpanKindValues.LLM.value:
        computed_data = build_llm_main_span_data(data)
    else:
        computed_data = {}

    attributes = {
        "source": data["source"],
        "traceId": data["traceId"],
        SpanAttributes.OPENINFERENCE_SPAN_KIND: main_span_kind,
        "beeai.version": data["version"],
        **computed_data,
    }

    with tracer.start_as_current_span(
        name="beeai-framework-main",
        start_time=data["startTime"],
        attributes=attributes,
    ) as current_span:
        run_error_span = next(
            (s for s in data["spans"] if s["attributes"]["target"] == data["run_error_span_key"]),
            None,
        )
        if run_error_span is not None:
            current_span.set_status(trace.StatusCode.ERROR)
        else:
            current_span.set_status(trace.StatusCode.OK)

        build_spans_for_parent(tracer, data["spans"], data["traceId"], parent_id=None)
